contentcuration/contentcuration/frontend/shared/utils/helpers.spec.js

```diff
@@ -38,300 +32,3 @@ describe('swapElements', () => {
     expect(swapElements(arr, idx1, idx2)).toEqual(expected);
   });
 });
-
-describe('removeDuplicatePairs', () => {
-  it('removes indentical pairs from an array', () => {
-    expect(
-      removeDuplicatePairs([
-        ['food', 'chocolate'],
-        ['drink', 'milk'],
-        ['chocolate', 'food'],
-        ['food', 'chocolate'],
-      ])
-    ).toEqual([
-      ['drink', 'milk'],
-      ['chocolate', 'food'],
-      ['food', 'chocolate'],
-    ]);
-  });
-});
-
-describe('isSuccessor', () => {
```

Comment:
Gotcha - had assumed they were covered in the getter tests, will move them there.

I did briefly investigate doing the anti-cyclic checks on the backend in the course of this refactor, but I think it would have to be implemented with a recursive CTE that might quickly run into performance issues.

Also, while it might give a suboptimal experience to have cycles in the prerequisite graph, it won't actually break anything in Kolibri - only self-referential prerequisites are hugely problematic.

Reply:
I used [Depth First Search](https://www.cs.cornell.edu/courses/cs2112/2019fa/lectures/lecture.html?id=traversals) traversal in `isSuccessor` that operates on the whole chart. I am not sure how backend implementation would work or how it compares to the solution you mention but I guess there might be some good algorithms out there that we could use for inspiration. Might be an interesting task for @micahscopes maybe.

Reply:
Yes, the helper function I am using is also doing a depth first search by doing an early return in the case of finding a matching path through the tree.

For the backend, a recursive CTE could be devised that operated in a similar way.